sql/src/main/java/io/crate/metadata/sys/SysShardsTableInfo.java

```diff
@@ -218,12 +232,26 @@ public RowGranularity rowGranularity() {
      * Any shards that are not yet assigned to a node will have a NEGATIVE shard id (see {@link UnassignedShard}
      */
     @Override
-    public Routing getRouting(WhereClause whereClause, @Nullable String preference) {
+    public Routing getRouting(WhereClause whereClause, @Nullable String preference, @Nullable SessionContext context) {
         // TODO: filter on whereClause
         Map<String, Map<String, List<Integer>>> locations = new TreeMap<>();
         ClusterState state = service.state();
+        GroupShardsIterator groupShardsIterator;
         String[] concreteIndices = state.metaData().getConcreteAllIndices();
-        GroupShardsIterator groupShardsIterator = state.getRoutingTable().allAssignedShardsGrouped(concreteIndices, true, true);
+        User user = context != null ? context.user() : null;
+        if (user != null) {
+            List<String> userAccessibleIndices = new ArrayList<>(concreteIndices.length);
```

Comment:
So we would try to process the table fqn for each shard. Meaning we would compute the same table fqn for multiple shards (as multiple shards will belong to the same index). Is that really better? 
It seems to me that with this approach the only waste of space is a collection (`userAccessibleIndices`) as with what you're proposing there will be way more TableIdent and String instances created